pkg/distributor/distributor.go

```diff
@@ -562,23 +562,29 @@ func shardByAllLabels(userID string, labels []mimirpb.LabelAdapter) uint32 {
 }
 
 // Remove the label labelName from a slice of LabelAdapters if it exists.
-func removeLabel(labelName string, labels *[]mimirpb.LabelAdapter) {
+// Returns true if label was removed.
+func removeLabel(labelName string, labels *[]mimirpb.LabelAdapter) bool {
 	for i := 0; i < len(*labels); i++ {
 		pair := (*labels)[i]
 		if pair.Name == labelName {
 			*labels = append((*labels)[:i], (*labels)[i+1:]...)
-			return
+			return true
 		}
 	}
+	return false
 }
 
 // Remove labels with value=="" from a slice of LabelPairs, updating the slice in-place.
-func removeEmptyLabelValues(labels *[]mimirpb.LabelAdapter) {
+// Returns true if labels were modified.
+func removeEmptyLabelValues(labels *[]mimirpb.LabelAdapter) bool {
```

Comment:
Instead of needing to remember to call `.ClearUnmarshalData()` when making changes to the timeseries, could methods like these be moved to `PreallocTimeseries` and handle clearing the cached data automatically? Seems less prone to someone forgetting to do this somewhere.

Reply:
Yes, I like that idea.

Reply:
Implemented in https://github.com/grafana/mimir/pull/5137/commits/849f61389f2debf8c79da99e786fe2c78159790b.